src/Stratis.Bitcoin/P2P/Protocol/Behaviors/EnforcePeerVersionCheckBehavior.cs

```diff
@@ -0,0 +1,111 @@
+﻿using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Configuration;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.Utilities;
+using TracerAttributes;
+
+namespace Stratis.Bitcoin.P2P.Protocol.Behaviors
+{
+    /// <summary>
+    /// Sets the minimum supported client version <see cref="this.NodeSettings.MinProtocolVersion"> to <see cref="this.Network.Consensus.Options.EnforcedMinProtocolVersion">
+    /// based on the predefined block height <see cref="this.Network.Consensus.Options.EnforceMinProtocolVersionAtBlockHeight">.
+    /// Once the new minimum supported client version is changed all existing peer connections will be dropped upon the first received message from outdated client.
+    /// </summary>
+    public class EnforcePeerVersionCheckBehavior : NetworkPeerBehavior
+    {
+        /// <summary>An indexer that provides methods to query the best chain (the chain that is validated by the full consensus rules)</summary>
+        private readonly ChainIndexer chainIndexer;
+
+        /// <summary>User defined node settings.</summary>
+        private readonly NodeSettings nodeSettings;
+
+        /// <summary>Specification of the network the node runs on - regtest/testnet/mainnet.</summary>
+        private readonly Network network;
+
+        /// <summary>Logger factory usded while cloning the object.</summary>
+        private readonly ILoggerFactory loggerFactory;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>
+        /// Set to <c>true</c> if the attached peer callbacks have been registered and they should be unregistered,
+        /// <c>false</c> if the callbacks are not registered.
+        /// </summary>
+        private bool callbacksRegistered;
+
+        /// <summary>
+        /// Initializes an instance of the object for outbound network peers.
+        /// </summary>
+        /// <param name="chainIndexer">The chain of blocks.</param>
+        /// <param name="nodeSettings">User defined node settings.</param>
+        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public EnforcePeerVersionCheckBehavior(ChainIndexer chainIndexer,
+            NodeSettings nodeSettings,
+            Network network,
+            ILoggerFactory loggerFactory)
+        {
+            this.chainIndexer = chainIndexer;
+            this.nodeSettings = nodeSettings;
+            this.network = network;
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
+        }
+
+        private Task OnMessageReceivedAsync(INetworkPeer peer, IncomingMessage message)
```

Comment:
formatting doesnt follow the codestyle, exit logs are missing.

Reply:
Added exit logs